src/modules.rs

```diff
@@ -22,22 +22,25 @@ use syntax::parse::parser;
 /// If a file is used twice in a crate, it appears only once.
 pub fn list_files<'a>(krate: &'a ast::Crate,
                       codemap: &codemap::CodeMap)
-                      -> HashMap<PathBuf, &'a ast::Mod> {
-    let mut result = HashMap::new();
+                      -> Vec<(PathBuf, &'a ast::Mod)> {
+    let mut result = Vec::new();
+    let mut files = HashSet::new();
```

Comment:
It seems somewhat wasteful to use a hash set and a Vec here just to maintain order (iiuc). Could you use a btree instead? That should have deterministic ordering and similar performance.

It would be good to comment somewhere about this change - it is somewhat subtle and I fear that it might accidentally be undone in the future. Ideally, we'd have a unit test too.